### DISCUSSION TITLE
TF adjustments are mostly <1?

### DISCUSSION THREAD

**Discussion by NickCrews** [moj-analytical-services/splink#1440] [Q&A]
I have a comparison on address. This is my code, it should at least give you some sense, but don't try to look at it too hard:

```python
def location_comparison(
    *,
    street_col: str,
    city_col: str,
    state_col: str,
    zipcode_col: str,
    latitude_col: str,
    longitude_col: str,
) -> dict:
    perfect_match_level = and_(
        _exact_match_level(street_col),
        or_(
            _exact_match_level(city_col),
            _exact_match_level(zipcode_col),
        ),
    )
    return {
        "output_column_name": "location",
        "comparison_description": "location comparison",
        "comparison_levels": [
            perfect_match_level,
            _exact_match_level(street_col),
            or_(
                _exact_match_level(zipcode_col),
                and_(
                    _exact_match_level(city_col),
                    _exact_match_level(state_col),
                ),
            ),
            # We don't seem to encounter this case in the data, which gives us warnings
            # about unfitted parameters.
            # cl_and(
            #     [
            #         cll.damerau_levenshtein_level(city_col, 2),
            #         _exact_match_level(state_col),
            #     ]
            # ),
            cll.distance_in_km_level(latitude_col, longitude_col, 100, not_null=True),
            cll.distance_in_km_level(latitude_col, longitude_col, 500, not_null=True),
            {
                "label_for_charts": "All location info null",
                "sql_condition": f"""
                    ({city_col}_l IS NULL OR {city_col}_r IS NULL)
                    AND
                    ({state_col}_l IS NULL OR {state_col}_r IS NULL)
                    AND
                    ({zipcode_col}_l IS NULL OR {zipcode_col}_r IS NULL)
                    AND
                    (
                           {latitude_col}_l IS NULL
                        OR {latitude_col}_r IS NULL
                        OR {longitude_col}_l IS NULL
                        OR {longitude_col}_r IS NULL
                    )
                """,
                "is_null_level": True,
            },
            cll.else_level(),
        ],
```

Take a look at the `_exact_match_level(street_col)` level. This is simply `street_l = street_r`. Note that this is the only level where a TF adjustment is computed, since all the other levels depend on multiple columns.

Here's the weird thing: If I look at this TF column in my comparisons (only for the comparisons that matched this level, since the rest of the comparisons have a TF set to 1.0 and are irrelevant), the vast majority are <1:
<img width="255" alt="image" src="https://github.com/moj-analytical-services/splink/assets/10820686/f775e2ed-a706-411c-83e4-712300a44a40">

Is this expected? I would sort of expect the mean to be 1? Is it because all the comparisons where TF > 1 are getting placed into the `perfect_match_level` level? I worry this is incorrectly reducing the match probability on all records that match this level, and I'm thinking of turning of TF adjustment.

**Comment by RobinL** [comment]
On my phone so haven't looked in detail, but this might help explain what's going on:
https://github.com/moj-analytical-services/splink/pull/1226

**Reply by RobinL** [reply]
Also just to note down what tf adjustments actually do (partly as a reminder to myself):

-  We can estimate u values by random sampling
-  For exact_match columns only, this isn't actually necessary - you can derive both  the precise **average** u value, and the precise term-specific u value from the distribution of column values.  If 10% of values are John, there's a 10% chance of 'collision'
- For the **average** u values, we use random sampling, because it generalises to all types of comparison levels, not just exact match
- The tf adjustments are based on the precise term-specific u values from the distribution of colum values.
- The tf adjustment is the difference between this precise value, and the average u value calculated from random sampling

I'm not sure that necessarily helps explain what you're seeing.  I agree that, overall, the positive adjustments should balance out the negative adjustments.

Is it possible actual comparisons that exist post-blocking are not representative of all comparisons, and hence in practice you don't necessarily see this balance
